File: Car Project Web/src/models/Booking.ts
import mongoose, { Schema, HydratedDocument, models } from 'mongoose';

// Interface for booking data
export interface IBooking {
  userId: string;
  carId: mongoose.Types.ObjectId;
  status: 'pending' | 'confirmed' | 'upcoming' | 'active' | 'completed' | 'cancelled';
  pickupDate: Date;
  returnDate: Date;
  pickupLocation: string;
  returnLocation?: string;
  pickupTime: string;
  returnTime: string;
  rentalDays: number;
  dailyRate: number;
  totalAmount: number;
  paymentStatus: 'pending' | 'paid' | 'refunded' | 'failed';
  bookingDate: Date;
  notes?: string;
  driverAge?: string;
  additionalDriver?: boolean;
  insurance?: string;
  createdAt: Date;
  updatedAt: Date;
}

// Type alias for hydrated documents
export type BookingDocument = HydratedDocument<IBooking>;

const BookingSchema = new Schema<IBooking>({
  userId: {
    type: String,
    required: true,
    index: true
  },
  carId: {
    type: Schema.Types.ObjectId,
    ref: 'Car',
    required: true,
    index: true
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'upcoming', 'active', 'completed', 'cancelled'],
    default: 'pending',
    required: true,
    index: true
  },
  pickupDate: {
    type: Date,
    required: true,
    index: true
  },
  returnDate: {
    type: Date,
    required: true,
    index: true
  },
  pickupLocation: {
    type: String,
    required: true
  },
  returnLocation: {
    type: String
  },
  pickupTime: {
    type: String,
    required: true
  },
  returnTime: {
    type: String,
    required: true
  },
  rentalDays: {
    type: Number,
    required: true,
    min: 1
  },
  dailyRate: {
    type: Number,
    required: true,
    min: 0
  },
  totalAmount: {
    type: Number,
    required: true,
    min: 0
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'refunded', 'failed'],
    default: 'pending',
    required: true,
    index: true
  },
  bookingDate: {
    type: Date,
    default: Date.now,
    required: true,
    index: true
  },
  notes: {
    type: String
  },
  driverAge: {
    type: String
  },
  additionalDriver: {
    type: Boolean,
    default: false
  },
  insurance: {
    type: String
  }
}, {
  timestamps: true,
  collection: 'bookings'
});

// Compound indexes for efficient queries
BookingSchema.index({ userId: 1, status: 1 });
BookingSchema.index({ userId: 1, pickupDate: 1 });
BookingSchema.index({ carId: 1, pickupDate: 1, returnDate: 1 });

export const Booking = models.Booking || mongoose.model<IBooking>('Booking', BookingSchema);

File: Car project mobile/lib/models/booking.dart
class Booking {
  final String id;
  final String carId;
  final String userId;
  final DateTime pickupDate;
  final DateTime dropoffDate;
  final double totalPrice;
  final String status; // 'pending', 'confirmed', 'completed', 'cancelled'
  final DateTime createdAt;

  Booking({
    required this.id,
    required this.carId,
    required this.userId,
    required this.pickupDate,
    required this.dropoffDate,
    required this.totalPrice,
    this.status = 'confirmed',
    required this.createdAt,
  });

  int get days {
    return dropoffDate.difference(pickupDate).inDays;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'carId': carId,
      'userId': userId,
      'pickupDate': pickupDate.toIso8601String(),
      'dropoffDate': dropoffDate.toIso8601String(),
      'totalPrice': totalPrice,
      'status': status,
      'createdAt': createdAt.toIso8601String(),
    };
  }

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] as String,
      carId: json['carId'] as String,
      userId: json['userId'] as String,
      pickupDate: DateTime.parse(json['pickupDate'] as String),
      dropoffDate: DateTime.parse(json['dropoffDate'] as String),
      totalPrice: (json['totalPrice'] as num).toDouble(),
      status: json['status'] as String? ?? 'confirmed',
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  Booking copyWith({
    String? id,
    String? carId,
    String? userId,
    DateTime? pickupDate,
    DateTime? dropoffDate,
    double? totalPrice,
    String? status,
    DateTime? createdAt,
  }) {
    return Booking(
      id: id ?? this.id,
      carId: carId ?? this.carId,
      userId: userId ?? this.userId,
      pickupDate: pickupDate ?? this.pickupDate,
      dropoffDate: dropoffDate ?? this.dropoffDate,
      totalPrice: totalPrice ?? this.totalPrice,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

